src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,511 @@ namespace nbl::asset
-                    if (members.array[i].offset == (last.offset + last.size))
-                        last.size += members.array[i].size;
-                    else
-                        r->push_back({shaderStage, members.array[i].offset, members.array[i].size});
+                    descInfo.count = stageIntroBindingInfo.count.count;
+                    descInfo.isRuntimeSizedFlag = true;
                 }
-            }
-            introspection++;
-            r++;
-        }
-    }
+                else
+                {
+                    const auto count = stageIntroBindingInfo.count;
+                    if (count.countMode == SDescriptorArrayInfo::DESCRIPTOR_COUNT::SPEC_CONSTANT)
+                    {
+                        if (!specConstants)
+                            return false;
 
-    core::vector<SPushConstantRange> merged;
+                        const auto& specConstantFound = specConstants->find(count.specID);
+                        if (specConstantFound == specConstants->end())
+                            return false;
 
-    SPushConstantRange rngToPush; rngToPush.offset = 0u; rngToPush.size = 0u;
-    uint32_t stageFlags = 0u;
-    for (uint32_t i = 0u; i < ICPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE/sizeof(uint32_t); i += sizeof(uint32_t))
-    {
-        SPushConstantRange curr; curr.offset = i; curr.size = sizeof(uint32_t);
+                        descInfo.count = count.count;
+                    }
+                    else
+                    {
+                        descInfo.count = stageIntroBindingInfo.count.count;
+                    }
 
-        uint32_t tmpFlags = 0u;
-        for (uint32_t stg=0u; stg<infos.size(); stg++)
-            for (const SPushConstantRange& rng : ranges[stg])
-                if (curr.overlap(rng))
-                    tmpFlags |= rng.stageFlags;
+                    if (descInfo.count == 0u)
+                        return false;
 
-        if (!i)
-            stageFlags = tmpFlags;
+                    descInfo.isRuntimeSizedFlag = false;
+                }
+            }
+            else
+            {
+                descInfo.count = 1u;
+                descInfo.isRuntimeSizedFlag = false;
+            }
 
-        if (!tmpFlags)
-            continue;
-        if (tmpFlags == stageFlags)
-            rngToPush.size += sizeof(uint32_t);
-        else
-        {
-            rngToPush.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-            merged.push_back(rngToPush);
-            stageFlags = 0u;
+            const auto& pplnIntroDataFoundBinding = m_descriptorSetBindings[i].find(descInfo);
+            if (pplnIntroDataFoundBinding != m_descriptorSetBindings[i].end())
+            {
+                if (pplnIntroDataFoundBinding->type != stageIntroBindingInfo.type)
+                    return false;
+                descInfo.count = std::max(pplnIntroDataFoundBinding->count, descInfo.count);
+            }
```

Comment:
either you look up both in `m_descriptorSetBindings` AND `descriptorsToMerge` or you initialize `descriptorsToMerge` to be a copy of `m_descriptorSetBindings` as per https://github.com/Devsh-Graphics-Programming/Nabla/pull/677#discussion_r1572639296

Just add
```cpp
const auto& selfIntersectionFound = descriptorsToMerge[i].find(descInfo);
if (selfIntersectionFound!=descriptorsToMerge[i].end() && selfIntersectionFound->type!=stageIntroBindingInfo.type)
    return false;
```

Right now you don't check "self collisions" i.e. bindings that repeat with different types in the same SPIR-V module (which would make the module invalid ofc)